direct-messages/README.md

```diff
@@ -10,24 +10,74 @@ because the same message can simultaneously:
 - give the new person the info needed to access the group
 - reveal to the group that this action happened
 
-## Mapping `feed_id` to a `recp_key`
+## Computing the `recp_key`
 
 We define a shared key that the sender + recipient can both derive:
 
 ```js
-function computeDirectMessageKey (my_secret, your_public) {
+function computeDirectMessageKey (my_dh_secret, my_dh_public, your_dh_public, my_feed_tfk, your_feed_tfk) {
   var hash = 'SHA256'
-  var salt = SHA256("envelope-direct-messsage-shared-key-extract-salt")
-  var input_keying_material = scalarmult(my_secret, your_public)
-
-  return hkdf.Extract(hash, salt, input_keying_material)
+  var length = 32
+  var salt = SHA256("envelope-dm-v1-extract-salt")
+  var info = slp_encode(
+      ["envelope-ssb-dm-v1/key"]
+      .concat(sort([
+		tfk_encode(my_dh_public) || my_feed_tfk,
+		tfk_encode(your_dh_public) || your_feed_tfk
+  ])))
+  var input_keying_material = scalarmult(my_dh_secret, your_dh_public)
+
+  return hkdf(input_keying_material, length, {salt, info, hash})
```

Comment:
Here's my JS implementation. 

```js
const hash = 'SHA256'
const length = 32
const salt = SHA256('envelope-dm-v1-extract-salt')

function directMessageKey (my_dh_secret, my_dh_public, my_feed_tfk, your_dh_public, your_feed_tfk) {
  var input_keying_material = Buffer.alloc(na.crypto_scalarmult_BYTES)
  na.crypto_scalarmult(input_keying_material, my_dh_secret, your_dh_public)

  var info_context = Buffer.from('envelope-ssb-dm-v1/key', 'utf8')
  var info_keys = [
    Buffer.concat([my_dh_public, my_feed_tfk]),
    Buffer.concat([your_dh_public, your_feed_tfk])
  ].sort()
  var info = slp.encode([info_context, ...info_keys])

  return hkdf(input_keying_material, length, { salt, info, hash })
}

function SHA256 (input) {
  const hash = crypto.createHash('sha256')

  hash.update(input)
  return hash.digest()
}
```

Note I've changed the order of the args in the function to be all "Mine" then all "Yours

Reply:
Here's my more generalised take : 

```js
const hash = 'SHA256'
const length = 32
const salt = SHA256('envelope-dm-v1-extract-salt')

function directMessageKey (my_dh_secret, my_dh_public, my_feed_tfk, your_dh_public, your_feed_tfk) {
  var  input_keying_material = scalarmult(my_dh_secret, your_dh_public)

  var info_context = Buffer.from('envelope-ssb-dm-v1/key', 'utf8')
  var info_keys = sort([
    my_dh_public || my_feed_tfk,
    your_dh_public || your_feed_tfk
  ])
  var info = slp.encode([info_context, ...info_keys])

  return hkdf(input_keying_material, length, { salt, info, hash })
}
```

Notes:
- `(dh_secret, dh_public)` is some Diffie-Hellman compatible keypair to be used for encryption 
    - currently we take feed keys (`ed25519` signing keys) and convert these to keys compatible with diffie-hellman (dh) shared key encryption (`curve25119` keys)
    - in the future we plan to generate dh encryption keys seperately
- `feed_tfk` is the "id" of a feed, namely the public part of that feed's signing keypair, encoded in "type-format-key" format (see [TFK][TFK])
- `||` means Buffer concat
- `sort` means sort these 2 buffers bytewise so that the smallest is first
- `slp.encode` is "shallow length-prefixed encode" (see [SLP][SLP])

Reply:
discussed more here : https://hackmd.io/2P_TQowLQVOv7KJtFI-_Kw?edit